### PR TITLE
MSFT_audio_emitter implementation

### PR DESCRIPTION
**Pull request by najadojo**
This change implements the [MSFT_audio_emitter](https://github.com/najadojo/glTF/blob/8109e86a26c9210814f8d236589ce92a2f9fa8f8/extensions/2.0/Vendor/MSFT_audio_emitter/README.md) extension for the glTF loader.

### CODE REVIEW

**Comment by deltakosh** [comment]
We already have ufo on assets.babylonjs.com/ufo.glb. is it the same?

File: loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+                    _ArrayItem.Assign(this._emitters);
+                    for (const emitterIndex of extension.emitters) {
+                        const emitter = GLTFLoader._GetProperty(extensionContext, this._emitters, emitterIndex);
+                        promises.push(this._loadEmitterAsync(`#/extensions/${this.name}/emitter/${emitter._index}`, emitter).then(() => {
+                            if (node._babylonMesh) {
+                                for (const sound of emitter._babylonSounds) {
+                                    sound.attachToMesh(node._babylonMesh);
+                                    if (emitter.innerAngle != undefined || emitter.outerAngle != undefined) {
+                                        sound.setLocalDirectionToMesh(new Vector3(0, 0, 1));
+                                        sound.setDirectionalCone(2 * Tools.ToDegrees(emitter.innerAngle == undefined ? Math.PI : emitter.innerAngle),
```
Comment (bghgary) [inline loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts]: `emitter.innerAngle == undefined ? Math.PI : emitter.innerAngle` can be simplified to `emitter.innerAngle || Math.PI`

File: loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+                    return Promise.all(promises).then(() => {});
+                });
+            });
+        }
+
+        protected _loadAnimationAsync(context: string, animation: _ILoaderAnimation): Nullable<Promise<void>> { 
+            return this._loadExtensionAsync<_ILoaderAnimationEvents>(context, animation, (extensionContext, extension) => {
+                return this._loader._loadAnimationAsync(extensionContext, animation).then(() => {
+                    const promises = new Array<Promise<void>>();
+                    let babylonAnimationGroup = animation._babylonAnimationGroup;
```
Comment (bghgary) [inline loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts]: let -> const and you can just put the `!` here to avoid putting it everywhere else below

File: loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+
+                    _ArrayItem.Assign(extension.events);
+                    for (const event of extension.events) {
+                        promises.push(this._loadAnimationEventAsync(`${context}/extension/${NAME}/events/${event._index}`, context, animation, event, babylonAnimationGroup!));
+                    }
+
+                    return Promise.all(promises).then(() => {
+                        // Make sure all audio stops when the animation is terminated.
+                        const getCallback = (pause: Boolean) => {
```
Comment (bghgary) [inline loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts]: this is inefficient and IMO harder to read than if you just repeat the for loop

File: loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+                                    } else {
+                                        emitter._babylonData!.sound!.stop();
+                                    }
+                                }
+                            };
+                        }
+                        babylonAnimationGroup!.onAnimationGroupEndObservable.add(getCallback(false));
+                        babylonAnimationGroup!.onAnimationGroupPauseObservable.add(getCallback(true));
+    
+                        babylonAnimationGroup!.normalize();
```
Comment (bghgary) [inline loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts]: why do you need to normalize again?

File: loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+                    });
+                });
+            });
+        }
+
+        private _getEventAction(sound: WeightedSound, action: _AnimationEventAction, time: number, startOffset?: number): (currentFrame: number) => void {
+            return (currentFrame: number) => {
```
Comment (bghgary) [inline loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts]: it would be more efficient if you check the action outside of the lambda instead

File: loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts (hunk 1 above)
Comment (bghgary) [reply loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts]: similar for a few cases below...

File: loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+            const babylonAnimation = babylonAnimationGroup.targetedAnimations[0];
+            const emitterIndex = event.emitter;
+            const emitter = GLTFLoader._GetProperty(`#/extensions/${this.name}/emitter`, this._emitters, emitterIndex);
+            return this._loadEmitterAsync(context, emitter).then(()=> {
+                const sound = emitter._babylonData!.sound;
+                if (sound) {
+                    var babylonAnimationEvent = new AnimationEvent(event.time, this._getEventAction(sound, event.action, event.time, event.startOffset));
+                    babylonAnimation.animation.addEvent(babylonAnimationEvent);
+                }
+                return Promise.resolve();
```
Comment (bghgary) [inline loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts]: extraneous line

File: loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+            };
+        }
+
+        private _loadAnimationEventAsync(context: string, animationContext: string, animation: _ILoaderAnimation, event: _ILoaderAnimationEvent, babylonAnimationGroup: AnimationGroup): Promise<void> {
+            if (babylonAnimationGroup.targetedAnimations.length == 0) {
+                return Promise.resolve();
+            }
+            const babylonAnimation = babylonAnimationGroup.targetedAnimations[0];
+            const emitterIndex = event.emitter;
+            const emitter = GLTFLoader._GetProperty(`#/extensions/${this.name}/emitter`, this._emitters, emitterIndex);
```
Comment (bghgary) [inline loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts]: context is wrong here also

File: src/Audio/babylon.sound.ts [hunk 8]
```diff
@@ -397,10 +403,58 @@ module BABYLON {
...
+                    this._soundPanner.coneOuterAngle = this._coneOuterAngle;
+                }
+            }
+        }
+
         public setPosition(newPosition: Vector3) {
             this._position = newPosition;
 
-            if (Engine.audioEngine.canUseWebAudio && this.spatialSound && this._soundPanner) {
+            if (Engine.audioEngine.canUseWebAudio && this.spatialSound && this._soundPanner && !isNaN(this._position.x) && !isNaN(this._position.y) && !isNaN(this._position.z)) {
```
Comment (bghgary) [inline src/Audio/babylon.sound.ts]: why would position be NaN?

File: src/Audio/babylon.sound.ts [hunk 9]
```diff
@@ -576,10 +633,6 @@ module BABYLON {
             }
         }
 
-        public getVolume(): number {
```
Comment (bghgary) [inline src/Audio/babylon.sound.ts]: why was this removed?

File: tests/unit/babylon/src/Loading/babylon.sceneLoader.tests.ts [hunk 10]
```diff
@@ -493,6 +494,18 @@ describe('Babylon Scene Loader', function () {
...
 
+        it('Load UFO with MSFT_audio_emitter', () => {
+            const scene = new BABYLON.Scene(subject);
+            return BABYLON.SceneLoader.ImportMeshAsync(null, "/Playground/scenes/", "ufo.glb", scene).then(result => {
+                expect(result.meshes.length, "meshes.length").to.equal(scene.meshes.length);
+                expect(result.particleSystems.length, "particleSystems.length").to.equal(0);
+                expect(result.animationGroups.length, "animationGroups.length").to.equal(3);
+                expect(scene.soundTracks.length, "scene.soundTracks.length").to.equal(1);
+                expect(scene.soundTracks[0].soundCollection.length, "scene.soundTracks[0].soundCollection.length").to.equal(3);
+                expect(scene.soundTracks[0].soundCollection[0].onEndedObservable.hasObservers(), "scene.soundTracks[0].soundCollection[0].onEndedObservable.hasObservers()").to.equal(true);
```
Comment (bghgary) [inline tests/unit/babylon/src/Loading/babylon.sceneLoader.tests.ts]: to.be.true

File: tests/unit/babylon/src/Loading/babylon.sceneLoader.tests.ts [hunk 11]
```diff
@@ -16,6 +16,7 @@ describe('Babylon Scene Loader', function () {
             .load(function () {
                 // Force apply promise polyfill for consistent behavior between PhantomJS, IE11, and other browsers.
                 BABYLON.PromisePolyfill.Apply(true);
+                BABYLON.Engine.audioEngine = new BABYLON.AudioEngine();
```
Comment (bghgary) [inline tests/unit/babylon/src/Loading/babylon.sceneLoader.tests.ts]: @deltakosh Does this look okay? Not sure what audio engine would mean for a null engine.

File: loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts [hunk 12]
```diff
@@ -0,0 +1,459 @@
...
+
+        /**
+         * Stop playback
+         */
+        public stop() {
+            this.isPlaying = false;
+            if (this._currentIndex !== undefined) {
+                this._sounds[this._currentIndex].stop();
+            }
+        }
```
[the remark this answers is not in the document]
Comment (najadojo) [reply loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts]: Could but there isn't a need.

File: loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts [hunk 13]
```diff
@@ -0,0 +1,459 @@
...
+
+        private _loadEmitterAsync(context: string, emitter: _ILoaderEmitter): Promise<void> {
+            emitter._babylonSounds = emitter._babylonSounds || [];
+            if (!emitter._babylonData) {
+                const clipPromises = new Array<Promise<void>>();
+                const name = emitter.name || `emitter${emitter._index}`;
+                const options = {
+                    loop: false,
+                    autoplay: false,
+                    volume: emitter.volume == undefined ? 1 : emitter.volume,
```
[the remark this answers is not in the document]
Comment (najadojo) [reply loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts]: No if the volume comes in as 0 we don't want it to be set to 1.

File: loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts (hunk 1 above)
Comment (najadojo) [reply loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts]: again if the value is 0 we don't want that reset to PI

File: loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts (hunk 3 above)
Comment (najadojo) [reply loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts]: why is it inefficient? I need to capture the current value of the emitter list, if I don't later when these care called it will be empty.

Comment (najadojo) [reply loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts]: oh you mean testing the conditional each loop iteration

File: src/Audio/babylon.sound.ts (hunk 8 above)
Comment (najadojo) [reply src/Audio/babylon.sound.ts]: This happens because the animation can start before the sound is loaded.

File: loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts (hunk 3 above)
Comment (bghgary) [reply loaders/src/glTF/2.0/Extensions/MSFT_audio_emitter.ts]: might as well not bother with this function and just inline it?

File: tests/unit/babylon/src/Loading/babylon.sceneLoader.tests.ts [hunk 14]
```diff
@@ -502,7 +504,7 @@ describe('Babylon Scene Loader', function () {
                 expect(result.animationGroups.length, "animationGroups.length").to.equal(3);
                 expect(scene.soundTracks.length, "scene.soundTracks.length").to.equal(1);
                 expect(scene.soundTracks[0].soundCollection.length, "scene.soundTracks[0].soundCollection.length").to.equal(3);
-                expect(scene.soundTracks[0].soundCollection[0].onEndedObservable.hasObservers(), "scene.soundTracks[0].soundCollection[0].onEndedObservable.hasObservers()").to.equal(true);
+                expect(scene.soundTracks[0].soundCollection[0].onEndedObservable.hasObservers(), "scene.soundTracks[0].soundCollection[0].onEndedObservable.hasObservers()").to.true;
```
Comment (bghgary) [inline tests/unit/babylon/src/Loading/babylon.sceneLoader.tests.ts]: nit: to.**be**.true

File: dist/preview release/what's new.md [hunk 15]
```diff
@@ -89,6 +90,7 @@
 - Added glTF loader settings to the GLTF tab in the debug layer ([bghgary](http://www.github.com/bghgary))
 - Added debug logging and performance counters ([bghgary](http://www.github.com/bghgary))
 - Added support for EXT_lights_imageBased ([bghgary](http://www.github.com/bghgary))
+- Added support for MSFT_audio_emitters ([najadojo](http://www.github.com/najadojo))
```
Comment (bghgary) [inline dist/preview release/what's new.md]: spelling: MSFT_audio_emitter

**Comment by deltakosh** [comment]
only one conflict and I can merge
